add %Y M %M and %M M %Y formats

davidtedfordholt opened the pull request:
used the structure of `yearquarter()` specification, addresses #142

davidtedfordholt committed:
add %Y M %M and %M M %Y formats

used the structure of `yearquarter()` specification, addresses #142

earowang commented on R/yearmonth.R:
```diff
@@ -65,8 +65,27 @@ yearmonth.Date <- function(x) {
-  assertDate(x)
-  new_yearmonth(anydate(x))
+  key_words <- regmatches(x, gregexpr("[[:alpha:]]+", x))
+  if (all(grepl("^[[:digit:]]{1~4}[[:space:]]*(m|mon|month)[[:space:]]*[[:digit:]]{1~4}$",
```
Can you assign the regular expression to a variable?
Is it too strict to include `[[:space:]]` here?

davidtedfordholt replied:
The reason to include `[[:space]]*` is to allow the possibility, since I've seen both `2018M01` and `2018 M01`. I haven't seen `2018M 1` before, but it's easy enough to allow for, and I can imagine `2018 month 1` in a spreadsheet (probably created by concatenation).

davidtedfordholt replied:
The initial reason I decided to go with such a strict format compared to the quarterly thing, honestly, is that `m` is a far more common letter than `q`. I'm far more worried about something that isn't a date but looks like one for the first few lines with `m` than `q`.

davidtedfordholt replied:
I've assigned the regular expression to a variable in a subsequent commit.

earowang replied:
Including `[[:space:]]*` doesn't seem to make any difference between `2018M01` and `2018M01`. Can you remove `[[:space:]]*`?

davidtedfordholt replied:
Done and done.

earowang commented on R/yearmonth.R:
```diff
@@ -65,8 +65,27 @@ yearmonth.Date <- function(x) {
+    yr_mon <- regmatches(x, gregexpr("[[:digit:]]+", x))
+    digits_lgl <- map_lgl(yr_mon, ~ !has_length(.x, 2))
+    digits_len <- map_int(yr_mon, ~ sum(nchar(.x)))
+    digits_ind <- nchar(flatten_chr(yr_mon))
```
What does this line exactly do? Why it can't be length of 3?

davidtedfordholt replied:
If we allow years before 1000AD using this format, there is no way to tell the difference between %y and %Y for 2-digit years. Also, it seems more likely that a 3-digit year specification is an error than that it refers to a year before 1000AD (especially using this format).

earowang replied:
I think it's okay to throw an error for 3-digit years for now. If there's a compelling use case using 3-digit years, we'll accomodate this later.

earowang reviewed (commented):
Thanks for helping with the issue.
Can you also include a couple of unit tests in the PR? Thanks.

davidtedfordholt committed:
add tests for %Ym%M yearmonth()

Creating a `yearmonth()` before January 1, 1000 AD will fail, if you use the %Ym%M format. I accept this shortcoming.

davidtedfordholt committed:
Merge pull request #1 from davidtedfordholt/patch-2

add tests for %Ym%M yearmonth()

davidtedfordholt committed:
save regular expression to variable

earowang commented on tests/testthat/test-yearmonth.R:
```diff
@@ -12,6 +12,17 @@ test_that("input types for yearmonth()", {
 })
 
+test_that("%Ym%M format for yearmonth()", {
+  expect_identical(yearmonth("2018 M01"), yearmonth("2018 Jan"))
```
This unit test fails on gh CI.

earowang commented on tests/testthat/test-yearmonth.R:
```diff
@@ -12,6 +12,17 @@ test_that("input types for yearmonth()", {
+  expect_identical(yearmonth("2018 M1"), yearmonth("2018 Jan"))
+  expect_identical(yearmonth("2018mon01"), yearmonth("2018 Jan"))
+  expect_identical(yearmonth("2018 month 12"), yearmonth("2018 Dec"))
+  expect_error(yearmonth("201M1"))
+  expect_error(yearmonth("2018M13"))
+  expect_error(yearmonth("201M811"))
```
Can you include the error msg you expect?

earowang commented on tests/testthat/test-yearmonth.R:
```diff
@@ -12,6 +12,17 @@ test_that("input types for yearmonth()", {
+  expect_error(yearmonth("201M1"))
+  expect_error(yearmonth("2018M13"))
+  expect_error(yearmonth("201M811"))
+  expect_error(yearmonth(c("2018M1", "2018 Feb", "2018M3")))
```
Actually I think they are valid inputs, although a mixture of styles. I'll leave comments above to allow for this.

earowang commented on R/yearmonth.R:
```diff
@@ -65,8 +65,27 @@ yearmonth.Date <- function(x) {
+    yr_lgl <- map(yr_mon, ~ grepl("[[:digit:]]{4}", .x))
+    yr <- as.integer(map2_chr(yr_mon, yr_lgl, ~ .x[.y]))
+    mon <- as.integer(map2_chr(yr_mon, yr_lgl, ~ .x[!.y]))
+    if (any(mon > 12)) {
+      abort("Months can't be greater than 12.")
+    }
+    yearmonth(12 * (yr - 1970) + mon - 1)
+  } else {
+    assertDate(x)
+    new_yearmonth(anydate(x))
+  }
```
```suggestion
    x <- paste(yr, mon)
  }
  assertDate(x)
  new_yearmonth(anydate(x))
```

earowang reviewed (commented):
Thank.

1. Can you add yourself as a contributor to the authors field?
2. Can you draft a news item that improves `yearmonth()`, including the issue number and your gh handle.

davidtedfordholt committed:
remove [[:space:]]* from regex

davidtedfordholt committed:
add ctb

davidtedfordholt committed:
add news for %Y M%M
